### ISSUE TITLE
tf.while_loop hangs when Dropout layer is used

### ISSUE THREAD

**Issue by gongzhitaao** [fchollet/keras#5469]
Simply put, `tf.while_loop` hangs when `Dropout` layer is used.  Keras version `2.0.2` <s>`1.2.2`</s>, Tensorflow version `1.0.1` <s>`1.0.0`</s>.

Please see the [gist](https://gist.github.com/gongzhitaao/e56ba03607eccafc616ca9677f8d9ba6) to reproduce this error.

1. Basically If you remove (comment out) the `Dropout(0.5)` layer **OR** change the total iteration to less than 10, the program correctly runs and terminates.
2. Otherwise, the program hangs.

It seems to me like a race condition.  And it seems to be a bug within keras since tensorflow alone does not  have this bug.

**Comment by gongzhitaao** [comment]
This bug still persists...I will print my code here
```python
import os
# supress tensorflow logging other than errors
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'

import numpy as np
import tensorflow as tf

keras = tf.keras
from keras import backend as K
from keras.models import Sequential
from keras.layers import Dense, Dropout

print(keras.__version__)
print(tf.__version__)

sess = tf.InteractiveSession()
K.set_session(sess)

model = Sequential([
    Dense(10, input_shape=[3,]),
#    Dropout(0.5),               # comment out this line and it works
    Dense(1)])
model.compile(optimizer='adam', loss='binary_crossentropy',
              metrics=['accuracy'])

def _cond(xx, i):
    return tf.less(i, 20)

def _body(xx, i):
    y = model(xx)
    grad, = tf.gradients(y, xx)
    grad = tf.Print(grad, [grad])
    xx += grad
    return xx, i+1

x = tf.placeholder(tf.float32, (None, 3))
y, _ = tf.while_loop(_cond, _body, [x, 0], back_prop=False)

sess.run(tf.global_variables_initializer())

K.set_learning_phase(True)
feed_dict = {x: np.random.random((1, 3))}
val = sess.run(y, feed_dict=feed_dict)
print(val)
```
